src/feasibility_checker.jl

```diff
@@ -0,0 +1,141 @@
+#  Copyright 2017, Iain Dunning, Joey Huchette, Miles Lubin, and contributors
+#  This Source Code Form is subject to the terms of the Mozilla Public
+#  License, v. 2.0. If a copy of the MPL was not distributed with this
+#  file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+using LinearAlgebra
+
+"""
+    primal_feasibility_report(
+        model::Model,
+        [point::Dict{VariableRef,Float64}];
+        atol::Float64 = 1e-8,
+    )::Union{Nothing,Dict{Any,Float64}}
+
+Check the primal feasibility of `model` at the point given by the dictionary
+`point` mapping variables to primal values. If a variable is not given in
+`point`, the value is assumed to be `0.0`.
+
+If the point is feasible, this function returns `nothing`. If infeasible, this
+function returns a dictionary mapping the constriant reference of each violated
+constraint to the distance it is from being feasible.
+
+To obtain `point` from a solution of a solved model, use:
+```julia
+point = Dict(v => value(v) for v in all_variables(model))
+```
+"""
+function primal_feasibility_report(
+    model::Model,
+    point::Dict{VariableRef,Float64};
+    atol::Float64 = 1e-8,
+)
+    point_f = x -> get(point, x, 0.0)
+    violated_constraints = Dict{Any,Float64}()
+    for (F, S) in list_of_constraint_types(model)
+        # This loop is not type-stable because `F` and `S` change, but it should
+        # be fine because no one should be using this in performance-critical
+        # code.
+        for con in all_constraints(model, F, S)
+            obj = constraint_object(con)
+            d = _distance_to_set(value.(obj.func, point_f), obj.set)
```

Comment:
I'd merge these two into a function:
`constraint_feasibility(con, point_f; atol)`

Reply:
At one point I had exactly `_distance_to_set(::ConstraintRef, point_f::Function)`, but I went with this for simplicity. We can refactor it out once we agree on the high-level API.